internal/app/machined/pkg/runtime/v1alpha1/platform/metal/metal.go

```diff
@@ -64,28 +72,77 @@ func (m *Metal) Configuration(ctx context.Context) ([]byte, error) {
 }
 
 // PopulateURLParameters fills in empty parameters in the download URL.
-func PopulateURLParameters(downloadURL string, getSystemUUID func() (string, error)) (string, error) {
-	u, err := url.Parse(downloadURL)
+//nolint:gocyclo
+func PopulateURLParameters(ctx context.Context, downloadURL string, r state.State) (string, error) {
+	populatedURL := downloadURL
+
+	keyToVar := func(key string) string {
+		return `${` + key + `}`
+	}
+
+	genErr := func(varOfKey string, errToWrap error) error {
+		return fmt.Errorf("error while substituting %s: %w", varOfKey, errToWrap)
+	}
+
+	u, err := url.Parse(populatedURL)
 	if err != nil {
-		return "", fmt.Errorf("failed to parse %s: %w", constants.KernelParamConfig, err)
+		return "", fmt.Errorf("failed to parse %s: %w", populatedURL, err)
 	}
 
 	values := u.Query()
 
+	substitute := func(key string, getFunc func(ctx context.Context, r state.State) (string, error)) error {
+		for qKey, qValues := range values {
+			if len(qValues) > 0 {
+				qVal := qValues[0]
+
+				varOfKey := keyToVar(key)
+				varOfKeyExpr := regexp.QuoteMeta(varOfKey)
+				matcher := regexp.MustCompile(`(?i)` + varOfKeyExpr)
+
+				if matcher.MatchString(qVal) {
+					val, err := getFunc(context.Background(), r)
+					if err != nil {
+						return genErr(varOfKey, err)
+					}
+
+					qVal = matcher.ReplaceAllLiteralString(qVal, val)
+
+					values.Set(qKey, qVal)
+				}
+			}
+		}
+
+		return nil
+	}
+
+	if err := substitute(constants.UUIDKey, getSystemUUID); err != nil {
+		return "", err
+	}
+
+	if err := substitute(constants.SerialNumberKey, getSystemSerialNumber); err != nil {
+		return "", err
+	}
+
+	if err := substitute(constants.MacKey, getMACAddress); err != nil {
+		return "", err
+	}
+
+	if err := substitute(constants.HostnameKey, getHostname); err != nil {
+		return "", err
+	}
+
+	// Although the UUID can be substituted via the ${uuid} variable, we keep the behavior that the empty uuid= parameter is filled in with the UUID. This is backwards compatible.
 	for key, qValues := range values {
-		switch key {
-		case "uuid":
-			// don't touch uuid field if it already has some value
+		if key == constants.UUIDKey { // don't touch uuid field if it already has some value
```

Comment:
I keep thinking we can handle this in the function above as a special case for `uuid=` - if it exists and empty, treat it as `uuid=${UUID}`. it looks strange we tread it different way?